indykite/resource_ingest_pipeline.go

```diff
@@ -0,0 +1,159 @@
+		DeleteContext: configDeleteContextFunc(),
+		Importer: &schema.ResourceImporter{
+			StateContext: basicStateImporter,
+		},
+
+		Timeouts: defaultTimeouts(),
+		Schema: map[string]*schema.Schema{
+			locationKey:   locationSchema(),
+			customerIDKey: setComputed(customerIDSchema()),
+			appSpaceIDKey: setComputed(appSpaceIDSchema()),
+
+			nameKey:        nameSchema(),
+			displayNameKey: displayNameSchema(),
+			descriptionKey: descriptionSchema(),
+			createTimeKey:  createTimeSchema(),
+			updateTimeKey:  updateTimeSchema(),
+
+			ingestPipelineSourcesTypeKey: {
+				Type:        schema.TypeList,
+				Description: "List of sources to be used in the ingest pipeline.",
+				Required:    true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				MinItems: 1,
+				MaxItems: 10,
+			},
+			ingestPipelineOperationsTypeKey: {
+				Type:        schema.TypeList,
+				Description: "List of operations which will be allowed to be used in the ingest pipeline.",
+				Required:    true,
+				Elem: &schema.Schema{
+					Type:         schema.TypeString,
+					ValidateFunc: validation.StringInSlice(ingestPipelineAllowedOperations, false),
+				},
+				MinItems: 1,
+				MaxItems: 6,
+			},
+			ingestPipelineAppAgentTokenTypeKey: {
+				Type:        schema.TypeString,
+				Description: "Application agent token is used to identify the application space in IndyKite APIs.",
+				Required:    true,
+				ValidateFunc: validation.StringMatch(
+					ingestPipelineAppAgentTokenRegex, "must be valid application agent token.",
+				),
+			},
+		},
+	}
+}
+
+func resourceIngestPipelineFlatten(
+	data *schema.ResourceData,
+	resp *configpb.ReadConfigNodeResponse,
+) diag.Diagnostics {
+	var d diag.Diagnostics
+	ipCfg := resp.GetConfigNode().GetIngestPipelineConfig()
+	setData(&d, data, ingestPipelineAppAgentTokenTypeKey, ipCfg.GetAppAgentToken())
+
+	operationsMapping := make([]string, len(ipCfg.GetOperations()))
+	for i, intOp := range ipCfg.GetOperations() {
+		strOp, ok := IngestPipelineOperationTypesReverse[intOp]
+		if !ok {
+			d = append(d, buildPluginError(fmt.Sprintf("unsupported Ingest Pipeline Operation: %d", intOp)))
+			return d
+		}
+		operationsMapping[i] = strOp
+	}
+	setData(&d, data, ingestPipelineOperationsTypeKey, operationsMapping)
+
+	sourcesMapping := make([]string, len(ipCfg.GetSources()))
+	for i, source := range ipCfg.GetSources() {
+		sourcesMapping[i] = source
+	}
+	setData(&d, data, ingestPipelineSourcesTypeKey, sourcesMapping)
+
+	return d
+}
+
+func resourceIngestPipelineBuild(
+	d *diag.Diagnostics,
+	data *schema.ResourceData,
+	_ *ClientContext,
+	builder *config.NodeRequest,
+) {
+	sources := data.Get(ingestPipelineSourcesTypeKey).([]any)
+	operations := data.Get(ingestPipelineOperationsTypeKey).([]any)
+	cfg := &configpb.IngestPipelineConfig{
+		AppAgentToken: data.Get(ingestPipelineAppAgentTokenTypeKey).(string),
+	}
+
+	cfg.Sources = rawArrayToTypedArray[string](sources)
```

Comment:
Again, out of curiosity :laughing: You are writing so many new lines and new variables always. I'm just wondering why? Do you follow some best practice I don't know? Does that have a reason, or is it just your habit? 

I'm trying to make it as small as possible, but it also must be easy to understand. 

```diff
-	sources := data.Get(ingestPipelineSourcesTypeKey).([]any)
	operations := data.Get(ingestPipelineOperationsTypeKey).([]any)
	cfg := &configpb.IngestPipelineConfig{
		AppAgentToken: data.Get(ingestPipelineAppAgentTokenTypeKey).(string),
+		Sources: rawArrayToTypedArray[string](data.Get(ingestPipelineSourcesTypeKey).([]any))
	}
-
-	cfg.Sources = rawArrayToTypedArray[string](sources)
```

Reply:
Yeah, it's not some kind of a best practice I follow 😁 just a habit or a style I have, but sometimes it leads to not really good looking code. And in the Golang domain it's more of an anti-pattern I guess...
I don't like the long and complex one-liners, so I try to be more verbose, but sometimes it's too much 🤷 

Fixed the code with half of the proposal 😁